src/server/plugins/engine/components/DatePartsField.test.ts

```diff
@@ -113,6 +118,178 @@ describe('Date parts field', () => {
 
     expect(conditonEvaluationStateValue).toBe('2024-12-31')
   })
+
+  describe('Form validation', () => {
+    const output = {
+      value: {
+        myComponent__day: 1,
+        myComponent__month: 1,
+        myComponent__year: 2001
+      }
+    }
+
+    describe.each([
+      {
+        description: 'Trim empty spaces',
+        component: {
+          title: 'Example date parts field',
+          name: 'myComponent',
+          type: ComponentType.DatePartsField,
+          options: {}
+        } satisfies DatePartsFieldComponent,
+        assertions: [
+          {
+            input: {
+              myComponent__day: ' 01',
+              myComponent__month: ' 01',
+              myComponent__year: ' 2001'
+            },
+            output
+          },
+          {
+            input: {
+              myComponent__day: '01 ',
+              myComponent__month: '01 ',
+              myComponent__year: '2001 '
+            },
+            output
+          },
+          {
+            input: {
+              myComponent__day: ' 01 \n\n',
+              myComponent__month: ' 01 \n\n',
+              myComponent__year: ' 2001 \n\n'
+            },
+            output
+          }
+        ]
+      }
+    ])('$description', ({ component: def, assertions }) => {
+      let component: DatePartsField
+
+      beforeEach(() => {
+        component = new DatePartsField(def, formModel)
+      })
+
+      it('validates empty value', () => {
+        const { formSchema } = component.children
+
+        const input = {
+          myComponent__day: '',
+          myComponent__month: '',
+          myComponent__year: ''
+        }
+
+        const output = {
+          value: {
+            myComponent__day: '',
+            myComponent__month: '',
+            myComponent__year: ''
+          },
+          error: new Error(
+            'day must be a number. month must be a number. year must be a number'
+          )
+        }
+
+        const result = formSchema.validate(input, opts)
+        expect(result).toEqual(output)
+      })
+
+      it.each([...assertions])(
+        'validates custom example',
+        ({ input, output }) => {
+          const { formSchema } = component.children
+
+          const result = formSchema.validate(input, opts)
+          expect(result).toEqual(output)
+        }
+      )
+    })
+  })
+
+  describe('State validation', () => {
+    const now = new Date()
+    const _1DayInPast = addDays(now, -1)
+    const _2DaysInPast = addDays(now, -2)
+    const _1DayInFuture = addDays(now, 1)
+    const _2DaysInFuture = addDays(now, 2)
```

Comment:
Can we put these inline?

Probably just as clear without the new naming convention

I do still prefer [`luxon`](https://moment.github.io/luxon/#/) `DateTime.now().minus({ days: 2 })` 😆

```suggestion
    const now = new Date()
```